Vagrantfile

```diff
@@ -12,36 +12,51 @@ else
 end
 settings = config_file
 
+domain = settings["domain"]
+
 Vagrant.configure("2") do |config|
   config.vm.box = settings["box"]
   config.vm.synced_folder "./", "/vagrant", type: "rsync", rsync__auto: true, rsync__exclude: ['files/*.tar.gz', 'files/*.tgz', 'collections/', 'group_vars/', 'logs/', 'roles/']
 
-  settings["hosts"].each do |host_data|
-    ip, hostname, domain, cpus, memory = *host_data
-    config.vm.define hostname, autostart: true do |cfg|
-      cfg.vm.hostname = hostname
-      cfg.vm.network "private_network", ip: ip
-
-      cfg.vm.provider :virtualbox do |vb|
-        vb.name = hostname # sets gui name for VM
-        vb.customize ["modifyvm", :id, "--memory", memory, "--cpus", cpus, "--hwvirtex", "on"]
-      end # end provider virtualbox
-
-      cfg.vm.provider :libvirt do |libvirt|
-        libvirt.cpus = cpus
-        libvirt.memory = memory
-      end # end provider libvirt
-
-    end # end define
-  end # end settings
+  settings["topology"].each do |(group, hosts)|
+    hosts.each do |host|
+      ip, hostname, cpus, memory = destructure_host_dict(host)
+      config.vm.define hostname, autostart: true do |cfg|
+        cfg.vm.hostname = hostname
+        cfg.vm.network "private_network", ip: ip
+  
+        cfg.vm.provider :virtualbox do |vb|
+          vb.name = hostname # sets gui name for VM
+          vb.customize ["modifyvm", :id, "--memory", memory, "--cpus", cpus, "--hwvirtex", "on"]
+        end # end provider virtualbox
+  
+        cfg.vm.provider :libvirt do |libvirt|
+          libvirt.cpus = cpus
+          libvirt.memory = memory
+        end # end provider libvirt
+  
+      end # end define
+    end # end hosts
+  end # end groups
 
   config.vm.provision :ansible do |ansible|
     ansible.playbook = "#{vagrantfile_dir}/ping.yml"
-    ansible.host_vars = settings["hosts"].each_with_object ({}) do |(ip, hostname, domain, cpus, memory), dict|
-      dict[hostname] = {
-        :ip     => ip,
-        :domain => domain,
-      }
+    ansible.host_vars = settings["topology"].each_with_object ({}) do |(group, hosts), hostvars_dict|
+      hosts.each_with_object (hostvars_dict) do |host, hostvar_dict|
+        ip, hostname, cpus, memory = destructure_host_dict(host)
+        hostvar_dict[hostname] = {
+          :ip     => ip,
+          :domain => domain,
+        }
+      end # end host_var
     end # end host_vars
+    ansible.groups = settings["topology"].each_with_object ({}) do |(group, hosts), groups|
```

Comment:
It is possible to generate an ansible group for each master ? Something like :
```ini
[master1]
master-01
[master2]
master-02
[master3]
master-03
```
For master only we need to put components on specific master.

Reply:
It's possible

Reply:
Because it could easily become "corrupted" to generate the master groups, I added an optional `groups` key to the host definition.

Reply:
Yes but it is weird to have 2 ways to specify groups.
Maybe something like
```yaml
topology:
  - cpus: 2
    hostname: edge-01
    ip: 192.168.56.10
    memory: 2048
    groups:
      - edge
  - cpus: 3
    hostname: master-01
    ip: 192.168.56.11
    memory: 3072
    groups:
      - master
      - master1
```

Reply:
That's right, I'll change that

Reply:
Done in last commit